File: Undergrad/TI0119 - PROCESSAMENTO DIGITAL DE SINAIS/FastFourierTransform/FFT.m
%%
x = [ones(1,6) zeros(1,58)];

X2 = fft(x);
X = fft_4(x, true);


figure;
subplot(2,2,1), stem(fftshift(abs(X)))
subplot(2,2,2), stem(fftshift(angle(X)))

subplot(2,2,3), stem(fftshift(abs(X2)))
subplot(2,2,4), stem(fftshift(angle(X2)))

function X = fft_4(x, inverse)
    % This is a radix-N FFT, using decimation in frequency
    N = length(x); M = log2(N)/(4/2);
 
    % Initialize variables for floating point sim
    X = complex(zeros(1,N));
    
    % FFT algorithm
    X_aux = x;
    for stage = 0:M-1
        for n=1:N/4
            X((1:4)+(n-1)*4) = radix_dif_bfly(X_aux(n:N/4:end), ... 
                                                  floor((n-1)/(4^stage)) *(4^stage), ... 
                                                  N, ...
                                                  inverse);
        end
        
        X_aux = X;
    end
    
    X = bitrevorder(X*N);
end

function Z = radix_dif_bfly(x, segment, N, inverse)
    A = (1/4) * x(1:4).';
    
    B = (exp(-1j*pi/2)) .^ ((0:4-1)' * (0:4-1)); 
    B = B(bitrevorder(1:4), :);
    
    if(inverse)
        W_aux = ((exp(2j*pi/N)) .^ (bitrevorder(0:4-1)*segment)).';
    else
        W_aux = ((exp(-2j*pi/N)) .^ (bitrevorder(0:4-1)*segment)).';
    end
    
    Z = (B * A) .* W_aux;
end

File: Undergrad/TI0119 - PROCESSAMENTO DIGITAL DE SINAIS/FastFourierTransform/FFT.py
import numpy as np

def fft(x_n, inverse=False):
	N = len(x_n); 
	W_N = np.exp((-1)**(not inverse) * 2j * np.pi / N)

	x_n = bitrevorder(x_n)

	size = 2
	while size <= N:
		step = size // 2
		for i in range(0, N, size):
			k = 0
			for j in range(i, i + step):
				z_k = (W_N**k) * x_n[j + step]

				x_n[j + step] = x_n[j] - z_k
				x_n[j] 		  = x_n[j] + z_k
				
				k += N // size
		size *= 2

	return (1/N)**inverse * x_n

def shift_fft(X_n):
	N = len(X_n)
	return np.hstack([np.flip(X_n[:(N//2)]), np.flip(X_n[(N//2):])])

def bitrevorder(x):
	N = len(x); size = np.log2(N).astype(int)
	return np.array([x[reverse_bin(i, size)] for i in range(N)], dtype=np.complex_)

def reverse_bin(num, size):
	binary = bin(num)
	reverse = binary[-1:1:-1] 
	reverse = reverse + (size - len(reverse))*'0'

	return int(reverse,2) 

def d_convolve(x, y):
	N = max(len(x), len(y))
	return [sum([x[m]*y[(n-m)%N] for m in range(0,N)]) for n in range(0,N)]

# Radix-4 Decimation-in-Time
def bitrevorder_4(x):
	N = len(x); size = np.log2(np.sqrt(N)).astype(int)
	return np.array([x[0:N-3:4], x[1:N-2:4], x[2:N-1:4], x[3:N:4]]).flatten()

def fft_4(x_n, inverse=False):
	N = len(x_n); 
	W_N = np.exp((-1)**(not inverse) * 2j * np.pi / N)

	x_n = bitrevorder_4(x_n)

	size = 4
	while size <= N:
		step = size // 4
		for i in range(0, N, size):
			k = 0
			for j in range(i, i + step):
				print(size,step,i)
				z1_k = (W_N**(2*k)) * x_n[j + 1*step]
				z2_k = (W_N**k) 	* x_n[j + 2*step]
				z3_k = (W_N**(3*k)) * x_n[j + 3*step]

				
				x_n[j + 2*step] = x_n[j] -1j* z1_k - z2_k +1j * z3_k
				x_n[j + 1*step] = x_n[j] -    z1_k + z2_k - z3_k
				x_n[j + 3*step] = x_n[j] +1j* z1_k - z2_k -1j * z3_k
				
				x_n[j] 		  	= x_n[j] +    z1_k + z2_k +     z3_k
				
				k += N // size
		size *= 4

	return (1/N)**inverse * x_n

